### PR TITLE
Aws tutorial

### PR DESCRIPTION
**Pull request by kris-nova**
Closes https://github.com/kubernetes/kops/issues/1018

Giving our most cherished piece of documentation a little bit of love - help me with grammar and typos please :)

CC @brandoncole @evildandelions @chrislovecnm @justinsb 

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1032)

### CODE REVIEW

File: docs/aws.md [hunk 1]
```diff
@@ -1,39 +1,112 @@
...
-* Run "kops update cluster" to build your cluster:
+From Source:
+
+```
+go get -d k8s.io/kops
+cd ${GOPATH}/src/k8s.io/kops/
+git checkout release
+make
+```
+
```
Comment (chrislovecnm) [inline docs/aws.md]: Second option should be install from a release.

File: docs/aws.md [hunk 2]
```diff
@@ -1,39 +1,112 @@
...
+See our [installation guide](build.md) for more information
+
+## Install kubectl
+
+It is a good idea to grab a fresh copy of `kubectl` now if you don't already have it.
```
Comment (chrislovecnm) [inline docs/aws.md]: Kubectl is required, so ensure you have it?

File: docs/aws.md [hunk 3]
```diff
@@ -1,39 +1,112 @@
...
+* [Installation Guide](http://kubernetes.io/docs/user-guide/prereqs/)
+
+```
+wget -O https://github.com/kubernetes/kubernetes/releases/download/v1.4.6/kubernetes.tar.gz
+sudo cp kubernetes/platforms/darwin/amd64/kubectl /usr/local/bin/kubectl
+```
+
+## Setup your environment
+
+1) Set up a DNS hosted zone in Route 53, e.g. `mydomain.com`, and set up the DNS nameservers as normal so that domains will resolve.  You can reuse an existing domain name (e.g. `mydomain.com`), or you can create a "child" hosted zone (e.g. `myclusters.mydomain.com`) if you want to isolate them.
```
Comment (chrislovecnm) [inline docs/aws.md]: Can we find a tutorial we can link on how to setup route53. This is the bane our our existence.

File: docs/aws.md [hunk 4]
```diff
@@ -1,39 +1,112 @@
...
+
+## Create your first cluster
+
+Before we create a cluster, we need to generate a `cluster config`
+
+In these examples we assume you have already exported `KOPS_STATE_STORE`, otherwise you will need to append each command with `--state s3://<mystatestorebucket>`
+
 ```
-${GOPATH}/bin/kops update cluster ${NAME} --yes
+kops create cluster --zones=us-east-1c ${NAME}
```
Comment (chrislovecnm) [inline docs/aws.md]: Let's do console highlighting and $ prompt? Also can u export $NAME?

File: docs/aws.md [hunk 5]
```diff
@@ -1,39 +1,112 @@
...
+kops edit cluster ${NAME}
+```
+
+This will open up the cluster manifest YAML file in your favorite text editor. Here is where a user can define very specific parts of their cluster. For now, lets leave this as default.
+
+
+Lets go ahead and create the cluster in AWS!
+
+```
+kops update cluster ${NAME} --yes
```
Comment (chrislovecnm) [inline docs/aws.md]: Do we want to have them run it without --yes?

File: docs/aws.md [hunk 6]
```diff
@@ -1,39 +1,112 @@
...
+## Accessing your cluster
+
+A friendly reminder that kops runs asynchronously, and it will take your cluster a few minutes to come online.
+
+Remember when you installed `kubectl` earlier? The configuration for your cluster was automatically generated and written to `~/.kube/config` for you!
+
+A simple Kubernetes API call can be used to check if the API is online and listening. Let's use `kubectl`
+
+```
+kubectl get nodes
```
Comment (chrislovecnm) [inline docs/aws.md]: Kops validate cluster ;( pretty please

File: docs/aws.md [hunk 7]
```diff
@@ -1,39 +1,112 @@
...
+```
+
```
Comment (chrislovecnm) [inline docs/aws.md]: We can have them do a kubectl install busybox or something. We should link k8s tutorial as well.

**Comment by justinsb** [comment]
We still don't talk about NS records, which is where everyone falls over....

**Comment by kris-nova** [comment]
@chrislovecnm - there is a legal section regarding the pic, we are clear to use it.

File: docs/aws.md (hunk 1 above)
Comment (kris-nova) [inline docs/aws.md]: This is an install from a release.. 

```
22 git checkout release
```

File: docs/aws.md (hunk 2 above)
Comment (kris-nova) [inline docs/aws.md]: Isn't it only required if they want to use it after the cluster is online?

File: docs/aws.md (hunk 3 above)
Comment (kris-nova) [inline docs/aws.md]: Happy to add one if someone finds one :) Otherwise we can PR later

File: docs/aws.md (hunk 6 above)
Comment (kris-nova) [inline docs/aws.md]: What command would you like to use?

File: docs/aws.md (hunk 5 above)
Comment (kris-nova) [inline docs/aws.md]: I think it's more important to show it off, and explain what it does, than to try to add it in later as a note without an example.. Most users should be smart enough to drop it if they need to.

File: docs/aws.md (hunk 4 above)
Comment (kris-nova) [inline docs/aws.md]: Fixed in 890aa6d6090666a604ba17e8445650d09a54fbf5

File: docs/aws.md (hunk 7 above)
Comment (kris-nova) [inline docs/aws.md]: Is that in scope here?

**Comment by kris-nova** [comment]
@chrislovecnm - I have a few open questions for you
@justinsb - Can you elaborate a bit more on the NS concerns? What all should we add?

**Comment by chrislovecnm** [comment]
/lgtm

---

Reviewed 2 of 4 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kops/1032#-:-KY51_w6CcUuDsYJcEVt:b-8ks5qa)*

**Comment by chrislovecnm** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[docs/aws.md, line 92 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kops/1032#-KXy16edej5w8FfH1dN3-r2-92:-KY51qjoN0eHQUGyr0WK:b-cyh2ry) ([raw file](https://github.com/kubernetes/kops/blob/62279b3674b7b453469587d435bac8e984059fd2/docs/aws.md#L92)):*
<details><summary><i>Previously, kris-nova (Kris Childress) wrote…</i></summary><blockquote>

What command would you like to use?
</blockquote></details>

`kops validate cluster`

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kops/1032)*

**Comment by chrislovecnm** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[docs/aws.md, line 94 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kops/1032#-KXy16edej5w8FfH1dN3-r2-94:-KY52pxqolhyC0XdkeYH:bb5d3b8) ([raw file](https://github.com/kubernetes/kops/blob/62279b3674b7b453469587d435bac8e984059fd2/docs/aws.md#L94)):*
<details><summary><i>Previously, kris-nova (Kris Childress) wrote…</i></summary><blockquote>

Is that in scope here?
</blockquote></details>

What about a link to http://kubernetes.io/docs/user-guide/simple-nginx/

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kops/1032)*

**Comment by kris-nova** [comment]
@justinsb @chrislovecnm How do we want this PR and the content in https://www.nivenly.com/k8s-aws-private-networking/ to play together? If at all..?

**Comment by kris-nova** [comment]
@k8s-bot aws e2e test this

**Comment by chrislovecnm** [comment]
Getting the content from the blog post into AWS.md would be awesome.

**Comment by kris-nova** [comment]
@chrislovecnm @justinsb 

NS Notes are in here
A lot of stuff from the blog post has been ported over

If no one has anything else - please merge :)

File: docs/aws.md (hunk 2 above)
Comment (mrwacky42) [inline docs/aws.md]: There's also `brew install kubernetes-cli` which gets you `kubectl` (and bash completion for it)

Comment (kris-nova) [inline docs/aws.md]: Good tip @mrwacky42 

Added

**Comment by kris-nova** [comment]
@k8s-bot aws e2e test this

File: docs/aws.md [hunk 8]
```diff
@@ -1,39 +1,213 @@
...
+brew update && brew install awscli
+aws configure
+aws iam list-users
+```
+
+We should now be able to pull a list of IAM users from the API, verifying that our credentials are working as expected.
+
+## Configure DNS
+
+We will now need to set up DNS for cluster, find one of the scenarios below (A,B,C) that match your situation.
```
Comment (andrewgdavis) [inline docs/aws.md]: what about scenario (e) where a domain has not been bought; but user just wants to try kops?

**Comment by kris-nova** [comment]
@k8s-bot aws e2e test this

File: docs/aws.md (hunk 8 above)
Comment (kris-nova) [inline docs/aws.md]: That is a good one - I don't think we support using clusters without a DNS name.. although I could be wrong - @justinsb help here?

**Comment by kris-nova** [comment]
@k8s-bot aws e2e test this

File: docs/aws.md [hunk 9]
```diff
@@ -1,39 +1,213 @@
...
 ```
-${GOPATH}/bin/kops update cluster ${NAME} --yes
+
+* [Kubernetes Latest Release](https://github.com/kubernetes/kubernetes/releases/latest)
+
+* [Installation Guide](http://kubernetes.io/docs/user-guide/prereqs/)
+
+
+```bash
+wget -O https://github.com/kubernetes/kubernetes/releases/download/v1.4.6/kubernetes.tar.gz
```
Comment (chrislovecnm) [inline docs/aws.md]: v1.4.7 is out :)

File: docs/aws.md [hunk 10]
```diff
@@ -1,39 +1,213 @@
...
+sudo cp kubernetes/platforms/darwin/amd64/kubectl /usr/local/bin/kubectl
```
Comment (chrislovecnm) [inline docs/aws.md]: and please don't have them d/l k8s for kubectl :P  Give me one second, and I will find the docs on the site.

File: docs/aws.md [hunk 11]
```diff
@@ -1,39 +1,213 @@
...
+
+### Setting up a kops IAM user
+
+
+In this example we will be using a dedicated IAM user to use with kops. This user will need basic API security credentials in order to use kops. Create the user and credentials using the AWS console. [More information](http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSGettingStartedGuide/AWSCredentials.html).
+
+Kubernetes kops uses the official AWS Go SDK, so all we need to do here is set up your system to use the official AWS supported methods of registering security credentials defined [here](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials). Here is an example using the aws command line tool to set up your security credentials.
+
+```bash
+brew update && brew install awscli
```
Comment (chrislovecnm) [inline docs/aws.md]: Linux?  Point them to the aws docs?  http://docs.aws.amazon.com/cli/latest/userguide/installing.html

File: docs/aws.md [hunk 12]
```diff
@@ -1,39 +1,213 @@
...
+aws configure
+aws iam list-users
```
Comment (chrislovecnm) [inline docs/aws.md]: I like

```console
$ aws iam list-users
```
Also, mention that then need to run configure, but don't so the output.  It is multi-line crap we don't want here.

File: docs/aws.md [hunk 13]
```diff
@@ -1,39 +1,213 @@
...
+```
+
+We should now be able to pull a list of IAM users from the API, verifying that our credentials are working as expected.
+
+## Configure DNS
+
+We will now need to set up DNS for cluster, find one of the scenarios below (A,B,C) that match your situation.
+
+### (A) Setting up DNS for your cluster, with amazon as your registrar
```
Comment (chrislovecnm) [inline docs/aws.md]: > amazon

Amazon Web Services (AWS)

File: docs/aws.md [hunk 14]
```diff
@@ -1,39 +1,213 @@
...
+
+If you bought your domain with AWS, then you should already have a hosted zone in Route53.
+
+If you plan on using your base domain, then no more work is needed. If you plan on using a subdomain to build your clusters on you will need to create a 2nd hosted zone in Route53.
+
+```bash
+ID=$(uuidgen) && aws route53 create-hosted-zone --name subdomain.kubernetes.com --caller-reference $ID
```
Comment (chrislovecnm) [inline docs/aws.md]: again I like showing a `$` for denoting that we are on a console, but your call

File: docs/aws.md [hunk 15]
```diff
@@ -1,39 +1,213 @@
...
+  - Create the subdomain
+
+```bash
+ID=$(uuidgen) && aws route53 create-hosted-zone --name subdomain.kubernetes.com --caller-reference $ID
+```
+
+  - Note your hosted zone ID
+
+```bash
+aws route53 list-hosted-zones | jq '.HostedZones[] | select(.Name=="subdomain.kubernetes.com") | .Id'
```
Comment (chrislovecnm) [inline docs/aws.md]: we nee jq for that example ... mention install https://github.com/stedolan/jq/wiki/Installation

File: docs/aws.md [hunk 16]
```diff
@@ -1,39 +1,213 @@
...
+```
+
+Note: You don’t have to use environmental variables here. You can always define the values using the –name and –state flags later.
+
+#### Form your create cluster command
+
+We will need to note which availability zones are available to us. In this example we will be deploying our cluster to the us-west-1 region.
+
+```bash
+aws ec2 describe-availability-zones --region us-west-1
```
Comment (chrislovecnm) [inline docs/aws.md]: Don't use us-west-1 .. use us-east-1 or us-west-2 ... more than 3 az

Comment (chrislovecnm) [inline docs/aws.md]: And you used us-west-2 below.

File: docs/aws.md [hunk 17]
```diff
@@ -1,39 +1,213 @@
...
+
+#### Cluster Configuration
+
+We now have created the underlying cluster configuration, lets take a look at every aspect that will define our cluster.
+
+```bash
+kops edit cluster ${NAME}
+```
+
+This will open up the cluster config (that is actually stored in the S3 bucket we created earlier!) in your favorite text editor. Here is where we can optionally really tweak our cluster for our use case. In this tutorial, we leave it default for now.
```
Comment (chrislovecnm) [inline docs/aws.md]: Do you want that it defaults to vi?  Use EDITOR to tweak?

File: docs/aws.md [hunk 18]
```diff
@@ -1,39 +1,213 @@
...
+## Accessing your cluster
+
+A friendly reminder that kops runs asynchronously, and it will take your cluster a few minutes to come online.
+
+Remember when you installed `kubectl` earlier? The configuration for your cluster was automatically generated and written to `~/.kube/config` for you!
+
+A simple Kubernetes API call can be used to check if the API is online and listening. Let's use `kubectl`
+
+```bash
+kubectl get nodes
```
Comment (chrislovecnm) [inline docs/aws.md]: `kops validate cluster`??  Then `kubectl -n kube-system get po`?

File: docs/aws.md (hunk 10 above)
Comment (chrislovecnm) [inline docs/aws.md]: You can link this: http://kubernetes.io/docs/user-guide/prereqs/

**Comment by kris-nova** [comment]
@k8s-bot aws e2e test this

**Comment by kris-nova** [comment]
@k8s-bot aws e2e test this

File: docs/aws.md (hunk 9 above)
Comment (chrislovecnm) [inline docs/aws.md]: Actually 1.5.1

**Comment by chrislovecnm** [comment]
LGTM
